Adopt the C++ spaceship operator in a few more places

cdumez opened the pull request:
#### 087241f0f367083ea6f453c487068f1959c07fe4
<pre>
Adopt the C++ spaceship operator in a few more places
<a href="https://bugs.webkit.org/show_bug.cgi?id=291172">https://bugs.webkit.org/show_bug.cgi?id=291172</a>

Reviewed by Darin Adler.

* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarkerRange::AXTextMarkerRange):
(WebCore::operator&lt;=&gt;):
(WebCore::AXTextMarker::wordRange const):
(WebCore::partialOrder): Deleted.
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::operator==):
(WebCore::operator&lt;=&gt;):
(WebCore::operator==): Deleted.
(WebCore::operator!=): Deleted.
(WebCore::operator&lt;): Deleted.
(WebCore::operator&gt;): Deleted.
(WebCore::operator&lt;=): Deleted.
(WebCore::operator&gt;=): Deleted.
* Source/WebCore/animation/WebAnimationTime.cpp:
(WebCore::operator&lt;=&gt;):
* Source/WebCore/animation/WebAnimationTime.h:
* Source/WebCore/dom/Position.cpp:
(WebCore::operator&lt;=&gt;):
(WebCore::documentOrder): Deleted.
* Source/WebCore/dom/Position.h:
(WebCore::operator&lt;): Deleted.
(WebCore::operator&gt;): Deleted.
(WebCore::operator&gt;=): Deleted.
(WebCore::operator&lt;=): Deleted.
* Source/WebCore/editing/VisiblePosition.cpp:
(WebCore::operator&lt;=&gt;):
(WebCore::documentOrder): Deleted.
* Source/WebCore/editing/VisiblePosition.h:
(WebCore::operator&lt;): Deleted.
(WebCore::operator&gt;): Deleted.
(WebCore::operator&lt;=): Deleted.
(WebCore::operator&gt;=): Deleted.
* Source/WebCore/platform/Decimal.cpp:
(WebCore::Decimal::operator!= const): Deleted.
(WebCore::Decimal::operator&lt; const): Deleted.
(WebCore::Decimal::operator&lt;= const): Deleted.
(WebCore::Decimal::operator&gt; const): Deleted.
(WebCore::Decimal::operator&gt;= const): Deleted.
* Source/WebCore/platform/Decimal.h:
(WebCore::operator==):
(WebCore::operator!=):
(WebCore::operator&lt;=&gt;):
(WebCore::Decimal::operator== const): Deleted.
* Source/WebCore/platform/network/DNS.h:
(WebCore::IPAddress::operator&lt;=&gt;):
(WebCore::IPAddress::operator==):
(WebCore::IPAddress::compare const): Deleted.
(WebCore::IPAddress::operator&lt; const): Deleted.
(WebCore::IPAddress::operator&gt; const): Deleted.
(WebCore::IPAddress::operator== const): Deleted.
* Source/WebCore/svg/animation/SMILTime.h:
(WebCore::operator&lt;=&gt;):
(WebCore::operator&gt;): Deleted.
(WebCore::operator&lt;): Deleted.
(WebCore::operator&gt;=): Deleted.
(WebCore::operator&lt;=): Deleted.
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::OperatingDate::operator&lt; const): Deleted.
(WebKit::OperatingDate::operator&lt;= const): Deleted.
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
(WebKit::OperatingDate::operator&lt;=&gt;):
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(WebKit::TrackerAddressLookupInfo::find):
* Tools/TestWebKitAPI/Tests/WebCore/IPAddressTests.cpp:
(TestWebKitAPI::TEST(IPAddressTests, CompareIPAddresses)):

Canonical link: <a href="https://commits.webkit.org/293372@main">https://commits.webkit.org/293372@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/154d8bb49b738e4d0c5cc0e3d73a36a5cc197370

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98703 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18336 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8573 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103828 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49291 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100748 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18629 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26787 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75152 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32300 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101707 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14152 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89139 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55509 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13937 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7104 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48673 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/91390 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83887 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7182 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106199 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/97332 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25793 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18808 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/84126 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26170 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85340 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83613 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21119 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28253 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5928 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19504 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25751 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30933 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/120950 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25569 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33840 "Found 19779 jsc stress test failures: ChakraCore.yaml/ChakraCore/test/Array/LastUsedSegmentHasNULLElement.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_ctr.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_fastinit.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_includes.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_indexOf.js.default ..., JSC test binary failure: testapi") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28889 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27144 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/network/DNS.h:
```diff
@@ -71,35 +71,29 @@ class IPAddress {
-    ComparisonResult compare(const IPAddress& other) const
+    friend std::partial_ordering operator<=>(const IPAddress& a, const IPAddress& b)
     {
         auto comparisonResult = [](int result) {
```
Could we do without this if we had a `compareSpans` function that returned `std::strong_ordering`?

cdumez replied:
I'll update `compareSpans` to return a `std::strong_ordering`

darinadler commented on Source/WebCore/platform/network/DNS.h:
```diff
@@ -71,35 +71,29 @@ class IPAddress {
-    bool operator==(const IPAddress& other) const { return compare(other) == ComparisonResult::Equal; }
+    friend bool operator==(const IPAddress& a, const IPAddress& b)
+    {
+        return (a <=> b) == std::partial_ordering::equivalent;
```
```suggestion
        return is_eq(a <=> b);
```

darinadler commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -569,16 +569,14 @@ static void populateIfNeeded()
-                switch (address.compare(list[middle].m_network)) {
-                case WebCore::IPAddress::ComparisonResult::Equal:
+                auto compareResult = address <=> list[middle].m_network;
+                if (compareResult == std::partial_ordering::equivalent)
```
```suggestion
            if (is_eq(compareResult))
```

darinadler commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -569,16 +569,14 @@ static void populateIfNeeded()
+                if (compareResult == std::partial_ordering::equivalent)
                     return &list[middle];
-                case WebCore::IPAddress::ComparisonResult::Less:
+                if (compareResult == std::partial_ordering::less)
```
```suggestion
            if (is_lt(compareResult))
```

darinadler commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -569,16 +569,14 @@ static void populateIfNeeded()
                     upper = middle;
-                    break;
-                case WebCore::IPAddress::ComparisonResult::Greater:
+                else if (compareResult == std::partial_ordering::greater)
```
```suggestion
            if (is_gt(compareResult))
```

cdumez committed:
Adopt the C++ spaceship operator in a few more places
https://bugs.webkit.org/show_bug.cgi?id=291172

Reviewed by Darin Adler.

* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarkerRange::AXTextMarkerRange):
(WebCore::operator<=>):
(WebCore::AXTextMarker::wordRange const):
(WebCore::partialOrder): Deleted.
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::operator==):
(WebCore::operator<=>):
(WebCore::operator==): Deleted.
(WebCore::operator!=): Deleted.
(WebCore::operator<): Deleted.
(WebCore::operator>): Deleted.
(WebCore::operator<=): Deleted.
(WebCore::operator>=): Deleted.
* Source/WebCore/animation/WebAnimationTime.cpp:
(WebCore::operator<=>):
* Source/WebCore/animation/WebAnimationTime.h:
* Source/WebCore/dom/Position.cpp:
(WebCore::operator<=>):
(WebCore::documentOrder): Deleted.
* Source/WebCore/dom/Position.h:
(WebCore::operator<): Deleted.
(WebCore::operator>): Deleted.
(WebCore::operator>=): Deleted.
(WebCore::operator<=): Deleted.
* Source/WebCore/editing/VisiblePosition.cpp:
(WebCore::operator<=>):
(WebCore::documentOrder): Deleted.
* Source/WebCore/editing/VisiblePosition.h:
(WebCore::operator<): Deleted.
(WebCore::operator>): Deleted.
(WebCore::operator<=): Deleted.
(WebCore::operator>=): Deleted.
* Source/WebCore/platform/Decimal.cpp:
(WebCore::Decimal::operator!= const): Deleted.
(WebCore::Decimal::operator< const): Deleted.
(WebCore::Decimal::operator<= const): Deleted.
(WebCore::Decimal::operator> const): Deleted.
(WebCore::Decimal::operator>= const): Deleted.
* Source/WebCore/platform/Decimal.h:
(WebCore::operator==):
(WebCore::operator!=):
(WebCore::operator<=>):
(WebCore::Decimal::operator== const): Deleted.
* Source/WebCore/platform/network/DNS.h:
(WebCore::IPAddress::operator<=>):
(WebCore::IPAddress::operator==):
(WebCore::IPAddress::compare const): Deleted.
(WebCore::IPAddress::operator< const): Deleted.
(WebCore::IPAddress::operator> const): Deleted.
(WebCore::IPAddress::operator== const): Deleted.
* Source/WebCore/svg/animation/SMILTime.h:
(WebCore::operator<=>):
(WebCore::operator>): Deleted.
(WebCore::operator<): Deleted.
(WebCore::operator>=): Deleted.
(WebCore::operator<=): Deleted.
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::OperatingDate::operator< const): Deleted.
(WebKit::OperatingDate::operator<= const): Deleted.
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
(WebKit::OperatingDate::operator<=>):
* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(WebKit::TrackerAddressLookupInfo::find):
* Tools/TestWebKitAPI/Tests/WebCore/IPAddressTests.cpp:
(TestWebKitAPI::TEST(IPAddressTests, CompareIPAddresses)):

Canonical link: https://commits.webkit.org/293372@main
